lib/events.js

```diff
@@ -189,38 +133,103 @@ EventEmitter.prototype.emit = function emit(type) {
     return false;
   }
 
-  handler = events[type];
+  fn = events[type];
 
-  if (!handler)
+  if (!fn)
     return false;
 
   if (domain && this !== process) {
     domain.enter();
     needDomainExit = true;
   }
 
-  var isFn = typeof handler === 'function';
-  len = arguments.length;
-  switch (len) {
-    // fast cases
-    case 1:
-      emitNone(handler, isFn, this);
-      break;
-    case 2:
-      emitOne(handler, isFn, this, arguments[1]);
-      break;
-    case 3:
-      emitTwo(handler, isFn, this, arguments[1], arguments[2]);
-      break;
-    case 4:
-      emitThree(handler, isFn, this, arguments[1], arguments[2], arguments[3]);
-      break;
-    // slower
-    default:
-      args = new Array(len - 1);
-      for (i = 1; i < len; i++)
-        args[i - 1] = arguments[i];
-      emitMany(handler, isFn, this, args);
+  var len = arguments.length;
+  var args;
+  if (typeof fn === 'function') {
+    switch (len) {
+      case 1: fn.call(this); break;
+      case 2: fn.call(this, arguments[1]); break;
+      case 3: fn.call(this, arguments[1], arguments[2]); break;
+      case 4: fn.call(this, arguments[1], arguments[2], arguments[3]);
+        break;
+      default:
+        args = new Array(len - 1);
+        for (i = 1; i < len; i++)
+          args[i - 1] = arguments[i];
+        fn.apply(this, args);
+    }
+  } else {
+    var fnlen = fn.length;
+    fn = arrayClone(fn, fnlen);
+    switch (len) {
+      case 1:
+        fn[0].call(this);
+        fn[1].call(this);
+        if (fnlen === 2) break;
+        fn[2].call(this);
+        if (fnlen === 3) break;
+        fn[3].call(this);
+        if (fnlen === 4) break;
+        fn[4].call(this);
+        if (fnlen === 5) break;
```

Comment:
> if (fnlen === 5) [](start = 7, length = 17)

Is there any significance of unrolling the loop for `5`  ? Or is it that in most common scenario there are not more than 5 event handlers to trigger? Either case, could you write a comment about it?